### PR TITLE
Killer tomatoes

### PR DESCRIPTION
**Pull request by TheShuEd**
## About the PR
add new mutation - killer tomatoes. it can be produced by a successful mutation of blood tomatoes. rather weak creatures that are only friendly to those who were around at the time of their appearance. Everyone else will be bitten
pose a great danger if the botanist raise an army of...

## Why / Balance
because it's fun. port ss13 content

## Technical details
add NPCImpritingBehaviourComponent, Allows you to make a creature friendly to those who were nearby at the time of spawning.

## Media

https://github.com/space-wizards/space-station-14/assets/96445749/4ecfa582-ac5e-454a-9925-a7c8cb24e481

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- add: botanist can now mutate blood tomatoes into killer tomatoes! they will be personal pets, aggressively defending the owner from any social interactions.

### CODE REVIEW

**Comment by TheShuEd** [comment]
And yes, you can keep mutating tomatoes by making them bigger or slippery :trollface:

**Comment by Tayrtahn** [comment]
I'm guessing that's meant to be "duck**l**ing syndrome"? I might go with ImprintingBehavior to better get the point across.

These sound like fun, and the requirement of a mutation of a mutation should keep from being too common of a sight.

(Sorry to keep picking on your spelling and word choice (I realize English isn't your first language) but I think it's important to phrase things well to make the code more understandable)

**Comment by TheShuEd** [comment]
that's interesting, it turns out the term even has a different meaning in English. I'm going to rename the system

**Comment by UbaserB** [comment]
Frontier already has these, lol.

**Comment by TheShuEd** [comment]
I didn't know that, and it doesn't matter. :)

**Comment by K-Dynamic** [comment]
These are good but mob swarms can steamroll very easily

Mobs pass through each other and don't miss even with 50 other sprites stacked on top

On the other hand, I want to see botanists unleash 50 killer tomatoes on bloodreds, so...

**Comment by beck-thompson** [comment]
Honestly I say fun first balance later. I really think this guys would be so fun!

**Comment by TheShuEd** [comment]
4 beute damage and 24 hp per tomato

File: Content.Server/NPC/Components/NPCImpritingBehaviourComponent.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+{
+    [DataField]
+    public HashSet<EntityUid> ImpritingTarget = new();
+
+    [DataField]
+    public float SearchRadius = 3f;
+
+    [DataField]
+    public bool Follow = true;
+}
```
Comment (metalgearsloth) [inline Content.Server/NPC/Components/NPCImpritingBehaviourComponent.cs:19]:
```suggestion
namespace Content.Server.NPC.Components;
/// <summary>
/// A component that makes the entity friendly to nearby creatures it sees on init.
/// </summary>
[RegisterComponent]
[Access(typeof(NPCImpritingBehaviourSystem))]
public sealed partial class NPCImpritingBehaviourComponent : Component
{
    [DataField]
    public HashSet<EntityUid> ImpritingTargets = new();

    [DataField]
    public float SearchRadius = 3f;

    [DataField]
    public bool Follow = true;
}
```

Comment (metalgearsloth) [reply Content.Server/NPC/Components/NPCImpritingBehaviourComponent.cs:19]: Also comments please.

File: Content.Server/NPC/Components/FactionExceptionComponent.cs [hunk 2]
```diff
@@ -6,7 +6,7 @@ namespace Content.Server.NPC.Components;
 /// Prevents an NPC from attacking ignored entities from enemy factions.
 /// Can be added to if pettable, see PettableFriendComponent.
 /// </summary>
-[RegisterComponent, Access(typeof(NpcFactionSystem))]
+[RegisterComponent, Access(typeof(NpcFactionSystem), typeof(NPCImpritingBehaviourSystem))]
 public sealed partial class FactionExceptionComponent : Component
 {
```
Comment (metalgearsloth) [inline Content.Server/NPC/Components/FactionExceptionComponent.cs:11]: If we start adding a billion access overrides they should be going through a system as then there's no reason to have access, but I'll fix this when I rework npcs.

File: Content.Server/NPC/Systems/NPCImpritingBehaviourSystem.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        if (impriting.Comp.Follow)
+        {
+            var mommy = _random.Pick(impriting.Comp.ImpritingTarget);
+            _npc.SetBlackboard(impriting, NPCBlackboard.FollowTarget, new EntityCoordinates(mommy, Vector2.Zero));
+        }
+
+        return;
+    }
```
Comment (metalgearsloth) [inline Content.Server/NPC/Systems/NPCImpritingBehaviourSystem.cs:42]: This should be using compregistry and not just checking for actors (which is pretty suss anyway).

File: Resources/Prototypes/Entities/Mobs/NPCs/miscellaneous.yml [hunk 4]
```diff
@@ -66,3 +66,97 @@
...
+  description: it seems today it's not you eating tomatoes, it's the tomatoes eating you.
+  components:
+  - type: Item
+    size: Huge
+  - type: NpcFactionMember
+    factions:
+      - SimpleHostile
+  - type: HTN
+    rootTask:
+      task: DragonCarpCompound
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Mobs/NPCs/miscellaneous.yml:88]: Just make a new compound please.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by K-Dynamic** [comment]
> 4 brute damage and 24 hp per tomato

Mutate slip trait, get CQC killer tomatoes :trollface:

File: Content.Server/NPC/Components/FactionExceptionComponent.cs (hunk 2 above)
Comment (TheShuEd) [reply Content.Server/NPC/Components/FactionExceptionComponent.cs:11]: added To Do

File: Resources/Prototypes/Entities/Mobs/NPCs/miscellaneous.yml [hunk 5]
```diff
@@ -85,8 +85,11 @@
       - SimpleHostile
   - type: HTN
     rootTask:
-      task: DragonCarpCompound
+      task: KillerTomatoCompound
   - type: NPCImpritingBehaviour
+    whitelist:
+      components:
+      - Actor
```
Comment (TheShuEd) [inline Resources/Prototypes/Entities/Mobs/NPCs/miscellaneous.yml:92]: @metalgearsloth I need to know a better criterion that selects only player-driven entities

Comment (TheShuEd) [reply Resources/Prototypes/Entities/Mobs/NPCs/miscellaneous.yml:92]: Not only humanoids

File: Content.Server/NPC/Systems/NPCImpritingBehaviourSystem.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            //if we haven't found mommy, we'll be aggressive with everyone.
+        }
+
+        if (impriting.Comp.Follow)
+        {
+            var mommy = _random.Pick(impritingTargets);
+            _npc.SetBlackboard(impriting, NPCBlackboard.FollowTarget, new EntityCoordinates(mommy, Vector2.Zero));
+        }
+
+        return;
```
Comment (ficcialfaint) [inline Content.Server/NPC/Systems/NPCImpritingBehaviourSystem.cs:42]:
```suggestion
```

**Review by ficcialfaint — COMMENTED** [review]
Isn't it **imprinting** and not **impriting**?

**Review by Tayrtahn — CHANGES_REQUESTED** [review]
> Isn't it **imprinting** and not **impriting**?

Yeah, that should be changed everywhere it shows up. Time for "rename symbol"!

**Comment by TheShuEd** [comment]
Shit
😨

**Comment by gitjubx** [comment]
> 4 brute damage and 24 hp per tomato

can understand the low health with the amount you can grow but does Hp scale with potency? would be neat the dmg seems be fine

**Comment by TheShuEd** [comment]
No, not scaling. I think its need botany refactor first

File: Resources/Prototypes/Entities/Mobs/NPCs/miscellaneous.yml (hunk 5 above)
Comment (metalgearsloth) [reply Resources/Prototypes/Entities/Mobs/NPCs/miscellaneous.yml:92]: humanoidappearance probably.

File: Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    private void OnMapInit(Entity<NPCImprintingOnSpawnBehaviourComponent> imprinting, ref MapInitEvent args)
+    {
+
+        var entities = _lookup.GetEntitiesInRange(imprinting, imprinting.Comp.SearchRadius);
+        var imprintingTargets = new List<EntityUid>();
+
+        if (entities == null) return;
+        if (entities.Count == 0) return;
```
Comment (metalgearsloth) [inline Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs:28]: 1. Entities can never be returned as null.
2. Use the overload that lets you pass in a set.

File: Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+        foreach (var ent in entities)
+        {
+            var check = imprinting.Comp.Whitelist != null && !imprinting.Comp.Whitelist.IsValid(ent);
+            if (!check)
```
Comment (metalgearsloth) [inline Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs:33]: Just wrap this into the if statement no reason to have it separate.

File: Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+            {
+                imprintingTargets.Add(ent);
+                var exception = EnsureComp<FactionExceptionComponent>(imprinting);
+                exception.Ignored.Add(ent);
+            }
```
Comment (metalgearsloth) [inline Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs:38]: Just make a helper method for this.

File: Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        base.Initialize();
+        SubscribeLocalEvent<NPCImprintingOnSpawnBehaviourComponent, MapInitEvent>(OnMapInit);
+    }
+
+    private void OnMapInit(Entity<NPCImprintingOnSpawnBehaviourComponent> imprinting, ref MapInitEvent args)
+    {
+
+        var entities = _lookup.GetEntitiesInRange(imprinting, imprinting.Comp.SearchRadius);
+        var imprintingTargets = new List<EntityUid>();
+
```
Comment (metalgearsloth) [inline Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs:26]: Use valuelist.

File: Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            var mommy = _random.Pick(imprintingTargets);
+            _npc.SetBlackboard(imprinting, NPCBlackboard.FollowTarget, new EntityCoordinates(mommy, Vector2.Zero));
+        }
+    }
+
+    private void AddImprintingTarget()
+    {
+
+    }
+}
```
Comment (metalgearsloth) [inline Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs:52]: Dummy method.

File: Content.Shared/NPC/Components/NPCImprintingOnSpawnBehaviourComponent.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    /// the radius in which all creatures will be marked as exceptions to attack
+    /// </summary>
+    [DataField]
+    public float SearchRadius = 3f;
+
+    /// <summary>
+    /// if there is a FollowCompound in HTN, the target of the following will be selected from random nearby targets when it appears
+    /// </summary>
+    [DataField]
+    public bool Follow = true;
```
Comment (metalgearsloth) [inline Content.Shared/NPC/Components/NPCImprintingOnSpawnBehaviourComponent.cs:27]: Should be on server.

File: Content.Shared/NPC/Components/NPCImprintingOnSpawnBehaviourComponent.cs [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+/// </summary>
+[RegisterComponent]
+public sealed partial class NPCImprintingOnSpawnBehaviourComponent : Component
+{
+    /// <summary>
+    /// filter who can be a friend to this creature
+    /// </summary>
+    [DataField]
+    public EntityWhitelist Whitelist;
+
```
Comment (metalgearsloth) [inline Content.Shared/NPC/Components/NPCImprintingOnSpawnBehaviourComponent.cs:16]: This should either be marked as required or nullable.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs (hunk 7 above)
Comment (TheShuEd) [reply Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs:28]: idk what is Overload

Comment (Tayrtahn) [reply Content.Server/NPC/Systems/NPCImprintingOnSpawnBehaviourSystem.cs]: An overload is a method (function) that has the same name as another method, but takes different parameters. Like `Foo(string bar)` could have overloads like `Foo(int baz)` or `Foo(string bar, bool baz)` etc.

There's an overload for GetEntitiesInRange that takes a hashset as a parameter.

File: Content.Server/NPC/Components/NPCImprintingOnSpawnBehaviourComponent.cs [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    /// filter who can be a friend to this creature
+    /// </summary>
+    [DataField]
+    public EntityWhitelist? Whitelist;
+
+    /// <summary>
+    /// the radius in which all creatures will be marked as exceptions to attack
+    /// </summary>
+    [DataField]
+    public float SearchRadius = 3f;
```
Comment (Partmedia) [inline Content.Server/NPC/Components/NPCImprintingOnSpawnBehaviourComponent.cs:21]: This seems like it duplicates the vision radius in the NPC blackboard

Comment (Partmedia) [reply Content.Server/NPC/Components/NPCImprintingOnSpawnBehaviourComponent.cs:21]: Or perhaps not, the initial impriting radius can be different from the final NPC search radius. Your variable name and comment made it hard for me to understand what this is for.

**Comment by TheShuEd** [comment]
wtf I didn't close the PR

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**